### PR TITLE
Fix randn_like inconsistency between eager and compile with fallback_random = True

### LINKED ISSUE #177652: [inductor] torch.compile's RNG state is inconsistent with eager on non-contiguous tensor when size >= 16
### 🐛 Describe the bug

When compiling a function with `torch.randn_like`, the compiled function's return result is inconsistent with eager even when the random seed is fixed and set `torch._inductor.config.fallback_random = True`.
This issue only happens when the input of the function is a non-contiguous tensor (created via `permute(1,0)`), which size larger than 16 and `shape[0], shape[1]!=1`.

Here is the code to reproduce:

```python
import torch
import itertools
torch._inductor.config.fallback_random = True

def fn(x):
    torch.manual_seed(0)
    return torch.randn_like(x)

# shape[0] * shape[1] >= 16 and shape[0], shape[1] != 1
for shape in itertools.product(range(1, 17), range(1, 17)):
    shape = shape
    x1 = torch.zeros(shape).permute(1, 0)
    x2 = torch.zeros(shape).permute(1, 0)
    
    cfunc = torch.compile(fn, backend='inductor')
    
    torch.manual_seed(0)
    out1 = fn(x1)
    # print(f'{out1=}')
    
    torch.manual_seed(0)
    out2 = fn(x1)
    # print(f'{out2=}')
    
    torch.manual_seed(0)
    out3 = cfunc(x2)
    # print(f'{out3=}')
    
    torch.manual_seed(0)
    out4 = cfunc(x2)
    # print(f'{out4=}')
    
    try:
        torch.testing.assert_close(out1, out3, equal_nan=True)
    except AssertionError as e:
        print(f'Shape {shape} will cause inconsistency.')

```

And the output shows that only specific shape can trigger this inconsistency:

```
Shape (2, 8) will cause inconsistency.
Shape (2, 9) will cause inconsistency.
Shape (2, 10) will cause inconsistency.
Shape (2, 11) will cause inconsistency.
Shape (2, 12) will cause inconsistency.
Shape (2, 13) will cause inconsistency.
Shape (2, 14) will cause inconsistency.
Shape (2, 15) will cause inconsistency.
Shape (2, 16) will cause inconsistency.
Shape (3, 6) will cause inconsistency.
Shape (3, 7) will cause inconsistency.
Shape (3, 8) will cause inconsistency.
Shape (3, 9) will cause inconsistency.
...
```

Note that this issue does not happen on backend `eager` and `aot_eager`.
This issue also does not happen when running on cuda (RTX 6000).

### Error logs

_No response_

### Versions

Versions of relevant libraries:
[pip3] numpy==2.4.2
[pip3] torch==2.12.0.dev20260316+cpu
[pip3] torchvision==0.26.0.dev20260223+cpu
[pip3] triton==3.6.0
[conda] numpy 2.4.2 pypi_0 pypi
[conda] torch 2.12.0.dev20260316+cpu pypi_0 pypi
[conda] torchvision 0.26.0.dev20260223+cpu pypi_0 pypi
[conda] triton 3.6.0 pypi_0 pypi

cc @chauhang @penguinwu

### PR DESCRIPTION
**Pull request by cleonard530**
Fixes #177652

When `fallback_random = True`, `torch.randn_like` was producing different results in eager vs compiled when the manual_seed was the same, the tensors where non-contiguous, and the total size >= 16. 

e.g.:
```
x = torch.zeros((6, 3)).t()

torch.manual_seed(0)
out_eager = torch.randn_like(x)

cfunc = torch.compile(torch.randn_like, backend='inductor')
torch.manual_seed(0)
out_compile = cfunc(x)
```
produced different results for `out_eager` and `out_compile`.

**Cause**
The reason for this is because the `normal_` generator for eager mode has a special fast path when the size of the tensor is >= 16 and the tensor is contiguous. In eager mode `torch.randn_like(x)` did not use this fast path since `x` was not contiguous, but the decomposition for inductor used `torch.randn` to generate a contiguous vector then transformed it after it was generated, so it did use the fast path. 

**Solution**
When `fallback_random=True`, remove `randn_like`, `rand_like`, and `randint_like` overloads from the copied decomposition table (avoid mutating the global decompositions dict). `rand_like` and `randint_like` were also removed to be consistent and because they used the same pattern as `randn_like` in the `decomposition.py` file.

Register ATen fallbacks in lowering.py for those op packets so graphs keep `aten.*_like` and execute the same kernels as eager.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @mlazos

### CODE REVIEW

File: torch/_inductor/decomposition.py [hunk 1]
```diff
@@ -963,7 +963,14 @@ def fast_random_decomps() -> dict[Any, Callable[..., Any]]:
...
     if config.fallback_random:
-        return decompositions
+        # Keep rand*_like as ATen fallbacks so RNG matches eager layout
+        # (see github.com/pytorch/pytorch/issues/177652).
+        table = dict(decompositions)  # make shallow copy
+        for name in ("rand_like", "randn_like", "randint_like"):
+            pkt = getattr(aten, name)
+            for ol in pkt.overloads():
+                table.pop(getattr(pkt, ol), None)
+        return table
```
Comment (jansel) [inline torch/_inductor/decomposition.py]: Can we update the decomp table directly and move these into `fast_random_decomps()` rather than patching the table on every call?

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 1 above)
Comment (cleonard530) [reply torch/_inductor/decomposition.py]: Changing the `decompositions` list directly could cause problems on later compiles if we change `fallback_random` from `True` to `False`. We still want the rank_like operations in the `decompositions` list when `fallback_random=False`. 

I could probably add a `fallback_random` flag to the `fast_random_decomps()` function (or create another cached function) so that I am not patching the table every time, but it would still need to use a copy of the `decompositons`. Is that similar to what you were thinking of?

Comment (jansel) [reply torch/_inductor/decomposition.py]: Isn't `fast_random_decomps` only used for `fallback_random=False`?  We already have two precomputed tables (one for fallback_random one without).

Comment (cleonard530) [reply torch/_inductor/decomposition.py]: (Can ignore, see my last comment)

Currently, the `fast_random_decomps()` isn't precomputed unless it's called first (i.e. call compile with `fallback_random=False`). So if we call `select_decomp_table()` with `fallback_random = True` and drop the `rand_like` operators, then call it with `fallback_random=False`, the `fast_random_decomps()` will not have `rand_like` operators in the return. 

One way to get around that could be to make sure `fast_random_decomps()` is called before we drop the `rand_like` methods. But the way it is setup now, that is not guaranteed. Do you want me to set it up so that `fast_random_decomps` is called first? Maybe with something like this:

```
@functools.cache
def fast_random_decomps_fallback_random() -> dict[Any, Callable[..., Any]]:
    fast_random_decomps()  # make sure this is called first
    for name in ("rand_like", "randn_like", "randint_like"):
        pkt = getattr(aten, name)
        for ol in pkt.overloads():
            decompositions.pop(getattr(pkt, ol), None)
    return decompositions 
```

Comment (cleonard530) [reply torch/_inductor/decomposition.py]: (Can ignore, see my last comment)

On second thought, if I used this, I would still probably use a copied table (`table = dict(decompositions)`). I don't think it would matter too much, but it seems safer to not update the global `decompositions` table and since it is cached I would only have to do it once.

Comment (cleonard530) [reply torch/_inductor/decomposition.py]: Ignore everything I have above. I now see what your mean by your first comment.  

Yes, we could add "rand_like", "randn_like", "randint_like" into the `fast_random_decomps()`. I will work on that and update the PR.

Sorry for the confusion.

File: torch/_inductor/decomposition.py [hunk 2]
```diff
@@ -775,22 +775,18 @@ def _rand_like(
     return result.permute(permutation).clone()
 
 
-@register_decomposition(aten.rand_like)
 def rand_like(self: torch.Tensor, **kwargs: Any) -> torch.Tensor:
```
Comment (jansel) [inline torch/_inductor/decomposition.py:780]: I think you can do something like:
```suggestion
@decomp.register_decomposition([aten.rand_like], extra_random_decomps)
def rand_like(self: torch.Tensor, **kwargs: Any) -> torch.Tensor:
```

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_inductor/decomposition.py (hunk 2 above)
Comment (cleonard530) [reply torch/_inductor/decomposition.py:780]: Thanks @jansel! Updated the file to use `@decomp.register_decomposition([aten.rand*_like], extra_random_decomps)` and removed the registrations inside the `fast_random_decomps()` function.

**Review by jansel — APPROVED** [review]

**Comment by cleonard530** [comment]
@jansel thank you for using the claude bot to describe the errors, it was very helpful. I updated the `inductor_fallback_ops` dict to include the `rand*_like` ops and regenerated the shims files and that seemed to get rid of the errors. 

I also removed the `xfail_if_mps` from the `test_like_rands*` test. The `test_like_rands_sliced_mps` failed on the last CI as well due to the `xfail_if_mps` and I believe the `test_like_rands_mps` would pass too due to these updates. Unfortunately, I don't have access to an mps device to check it myself right now. 

Le me know if you think this isn't the right approach or if there is anything else you want me to do/check. 

Thanks!
